include/lock/optimistic_lock.hpp

```diff
@@ -0,0 +1,287 @@
+   *
+   * NOTE: if a thread that does not have an exclusive lock calls this function, it will
+   * corrupt an internal lock status.
+   */
+  void
+  DowngradeToSIX()
+  {
+    lock_.fetch_add(kXLock + kSIXLock, std::memory_order_release);
+  }
+
+  /**
+   * @brief Release an exclusive lock.
+   *
+   */
+  void
+  UnlockX()
+  {
+    lock_.fetch_add(kXLock, std::memory_order_release);
+  }
+
+  /**
+   * @brief Get a shared lock with an intent-exclusive lock.
+   *
+   * NOTE: this function does not give up acquiring locks and continues spinlock.
+   */
+  void
+  LockSIX()
+  {
+    while (true) {
+      auto expected = lock_.load(std::memory_order_relaxed) & kXLockMask;
+      auto desired = expected + kSIXLock;
+      for (size_t i = 0; i < kRetryNum; ++i) {
+        const auto cas_success = lock_.compare_exchange_weak(
+            expected, desired, std::memory_order_acquire, std::memory_order_relaxed);
+        if (cas_success) return;
+
+        expected &= kXLockMask;
+        desired = expected + kSIXLock;
+        SPINLOCK_HINT
+      }
+
+      std::this_thread::sleep_for(kShortSleep);
+    }
+  }
+
+  /**
+   * @brief Upgrade an SIX lock to an X lock.
+   *
+   * NOTE: if a thread that does not have a shared lock with an intent-exclusive lock
+   * calls this function, it will corrupt an internal lock status.
+   */
+  void
+  UpgradeToX()
+  {
+    while (true) {
+      auto expected = (lock_ & kLockMask) + kSIXLock;
+      for (size_t i = 0; i < kRetryNum; ++i) {
+        const auto cas_success = lock_.compare_exchange_weak(
+            expected, kXLock, std::memory_order_acquire, std::memory_order_relaxed);
+        if (cas_success) return;
+
+        expected = (lock_ & kLockMask) + kSIXLock;
+        SPINLOCK_HINT
+      }
+
+      std::this_thread::sleep_for(kShortSleep);
+    }
+  }
+
+  /**
+   * @brief Release a shared lock with an intent-exclusive lock.
+   *
+   */
+  void
+  UnlockSIX()
+  {
+    lock_.fetch_sub(kSIXLock, std::memory_order_relaxed);
+  }
+
+ private:
+  /*####################################################################################
+   * Internal constants
+   *##################################################################################*/
+
+  /// a lock status for no locks.
+  static constexpr uint64_t kNoLocks = 0b000UL;
+
+  /// a lock status for shared locks.
+  static constexpr uint64_t kSLock = 0b001UL;
+
+  /// a lock status for exclusive locks.
+  static constexpr uint64_t kXLock = 0b010UL << 16;
+
+  /// a lock status for shared locks with intent-exclusive locks.
+  static constexpr uint64_t kSIXLock = 0b001UL << 16;
+
+  /// a bit mask for removing an X-lock flag.
+  static constexpr uint64_t kSLockMask = ~(0b010UL << 16);
+
+  /// a bit mask for removing SIX/X-lock flags.
+  static constexpr uint64_t kXLockMask = ~(0b011UL << 16);
+
+  /// a bit mask for removing S/SIX/X-lock flags.
+  static constexpr uint64_t kLockMask = ~((0b001 << 18) - 1);
```

Comment:
どちらでも同じですが，一応こっちの書き方の方がシンプルですかね（左シフトで生成される下位ビットって常にゼロであってましたっけ…？右シフトのときは最上位ビットに依存した気がするんですが）．

```suggestion
  static constexpr uint64_t kLockMask = (~0UL) << 18;
```

Reply:
確かにそうですね！
[ここ](https://docs.microsoft.com/ja-jp/cpp/cpp/left-shift-and-right-shift-operators-input-and-output?view=msvc-170#left-shifts)を読む限り，左シフトは符号なし数値では常に0で埋められると考えてよい気がします．